### DISCUSSION TITLE
How to select specific line in richtextbox in WPF?

### DISCUSSION THREAD

**Discussion by VahidEra** [dotnet/wpf#7812] [Q&A] [answered]
Hi Guys,
I have a richtextbox control in my WPF project that I want to select a specific line from.
I was able to do this to some extent with a lot of effort and research, but there is one problem here, and that is the line is not being selected completely. I do this by text, I would appreciate it if you could help me solve this problem And if there is another way to do this, I will be happy if you share it with me.
For better understanding and comfort, this is a sample of my project:
https://drive.google.com/file/d/1CndnLJ2aXNTo8c8I5U62pmsir-SQrbEe/view?usp=sharing

**Comment by miloush** [comment]
The problem is that [GetPositionAtOffset](https://learn.microsoft.com/en-us/dotnet/api/system.windows.documents.textpointer.getpositionatoffset?view=windowsdesktop-8.0) takes offset _in symbols_ while you are providing offset in characters. You are assuming that the string will stay uninterrupted, but that is not the case. For example, when you make the word "commodo" bold (by selecting it and pressing Ctrl+B) and then pressing your button, it will miss even more letters - there are more _symbols_ in the text to mark the formatted spans.

What are you actually trying to do? Highlight a line? Would it be different line with line-wrapping on and off? Or highlighting a specific text? If specific text, what if it is split across paragraph or other formatting? You might need to go through the runs one by one to find the one that your character offset falls in.

**Reply by VahidEra** [reply]
@miloush my dear friend Jan all I want is to change the color of a particular line only is there any way to do this because I am basically having a problem with richtextbox in WPF and I can't work with it properly if I can solve this the problem Everything will be solved

**Reply by miloush** [reply]
I am trying to figure out what does "line" mean to you. Is that a paragraph? If you resize a window and the text wraps, does the wrapped line count as a new line?

**Reply by miloush** [reply]
And how do you know which line you want? Is it a line number 4? Or is it a line containing word "commodo"?

**Reply by miloush** [reply, marked as answer]
https://learn.microsoft.com/en-us/dotnet/api/system.windows.documents.textpointer would be a useful reading to you. It has a simple example of how to find a string inside a text. You can call it in your demo like this:
```vb
    Public Sub SelectLine(ByVal text As String)
        Dim start As TextPointer = FindWordFromPosition(rtb.Document.ContentStart, text)
        Dim [end] As TextPointer = start.GetPositionAtOffset(text.Length)
    ...
```

But as I said, depending on where the text comes from and what users can do in the text box, this will break easily.

**Reply by VahidEra** [reply]
@miloush my dear friend Jan, The last code you posted for me here is exactly what I was looking for and solved my problem.
You got my point right. 
But I will give explanations about the questions you asked, which are all completely correct and professional. Because richtextbox in WPF is completely different from windows form, so I assumed that each row is a line. like this:
```
This is line 01: Lorem ipsum dolor sit amet, consectetuer adipiscing elit
This is Line 02: Aenean commodo ligula eget dolor
```
Yes, I know this is very silly :)))))) But it is enough for my purpose.
These two lines are separated by a line break. The program window will not be resized in any way And wherever a linebreak is inserted, I have considered it as a line, and it also helps me in counting the lines again Yes, I know this is very silly :)))))) .
According to your advice on my previous questions about RichTextBox: https://github.com/dotnet/wpf/discussions/7596 According to your guidance in my previous questions about richtextbox, which was for the same purpose, I managed to achieve my goal for this.
The goal was not to specify the word with a different color, but to specify the line with a different color, which I achieved thanks to your precise and beautiful instructions. It can be said that this is a trick about richtextbox in WPF.
You are the best, and the best never rest. Thank you very much for your constructive help and as always another problem was solved with your help.

Of course, I have used other codes for this purpose, which all together have helped this goal, for example, to count the lines and the length of the string, I even had to use unprincipled programming methods for ease of work, of course, it helped to solve the problem like this:
```
 Dim MyText As New TextRange(rtb.Document.ContentStart, rtb.Document.ContentEnd)

 Dim splittedLines() As String = MyText.Text.Split({Environment.NewLine}, StringSplitOptions.None) ' or StringSplitOptions.RemoveEmptyEntries

 For Each _line As String In splittedLines
          lis.Items.Add(_line )     'Here it helps me to get the string length for each line
 Next

 MessageBox.Show(splittedLines.Length.ToString())
